packages/token-bridge-sdk/src/hooks/useTokens.ts

```diff
@@ -143,6 +100,76 @@ function tokenListsToSearchableTokenStorage(
   )
 }
 
+export interface UseTokensParams {
+  l1ChainId: number
+  l2ChainId: number
+  bridgeTokens: ContractStorage<ERC20BridgeToken> | undefined
+}
+
+export interface UseTokensResult {
+  tokens: ContractStorage<ERC20BridgeToken> // all tokens - tokens from lists or users
+  tokensFromLists: ContractStorage<ERC20BridgeToken> // tokens fetched from lists
+  tokensFromUser: ContractStorage<ERC20BridgeToken> // tokens added by users
+  searchToken(tokenAddress: string): ERC20BridgeToken | null // search for a particular token
+}
+
+export function useTokens({
+  l1ChainId,
+  l2ChainId,
+  bridgeTokens
+}: UseTokensParams): UseTokensResult {
+  const tokenLists = useTokenLists(l2ChainId ? String(l2ChainId) : undefined)
+
+  /* tokens fetched from our standard token list */
+  const tokensFromLists = useMemo(() => {
+    if (typeof l1ChainId === 'undefined' || typeof l2ChainId === 'undefined') {
+      return {}
+    }
+
+    return tokenListsToSearchableTokenStorage(
+      tokenLists,
+      String(l1ChainId),
+      String(l2ChainId)
+    )
+  }, [tokenLists, l1ChainId, l2ChainId])
+
+  /* custom tokens added by the user, different from what are present in our approved lists */
+  const tokensFromUser = useMemo(() => {
+    const storage: ContractStorage<ERC20BridgeToken> = {}
+
+    // Can happen when switching networks.
+    if (typeof bridgeTokens === 'undefined') {
+      return {}
+    }
+
+    Object.keys(bridgeTokens).forEach((_address: string) => {
+      const bridgeToken = bridgeTokens[_address]
+
+      // Any tokens in the bridge that don't have a list id were added by the user.
+      if (bridgeToken && bridgeToken.listIds.size === 0) {
+        storage[_address] = { ...bridgeToken, listIds: new Set() }
+      }
+    })
+
+    return storage
+  }, [bridgeTokens])
+
+  /* All tokens - list or user */
+  const tokens = useMemo(() => {
+    return { ...tokensFromLists, ...tokensFromUser }
+  }, [tokensFromLists, tokensFromUser])
+
+  /*
+    function to search for a token within the union of token lists (pre-configured + user added)
+  */
+  const searchToken = (tokenAddress: string): ERC20BridgeToken | null => {
```

Comment:
I'm not sure we need this. We can use `tokens[tokenAddress]` directly where we call this hook. `tokens`, `tokensFromLists`, `tokensFromUser` are always defined, so in the worst case, it will return `undefined`.